src/Sarif/Baseline/ResultMatching/SarifLogResultMatcher.cs

```diff
@@ -268,23 +268,23 @@ private List<ExtractedResult> ExtractResultsFromRuns(IEnumerable<Run> sarifRuns)
                 indexRemappingVisitor.HistoricalLogicalLocations = resultPair.Run.LogicalLocations;
                 indexRemappingVisitor.VisitResult(result);
 
-                if (result.RuleIndex != -1)
+                string ruleId = result.ResolvedRuleId(resultPair.Run);
+                if (!string.IsNullOrEmpty(ruleId))
                 {
-                    ReportingDescriptor rule = result.GetRule(resultPair.Run);
-                    if (rule != null)
+                    if (reportingDescriptors.TryGetValue(ruleId, out int ruleIndex))
                     {
-                        if (reportingDescriptors.TryGetValue(rule, out int ruleIndex))
-                        {
-                            result.RuleIndex = ruleIndex;
-                        }
-                        else
-                        {
-                            result.RuleIndex = reportingDescriptors.Count;
-                            reportingDescriptors[rule] = reportingDescriptors.Count;
+                        result.RuleIndex = ruleIndex;
+                        result.RuleId = ruleId;
+                    }
+                    else
+                    {
```

Comment:
>{ [](start = 20, length = 1)

It looks like you don't set `result.RuleId` in the `else` branch. If so, a test hole as well.

Reply:
Yep - good catch, and added a test to cover Rule lookup by index and id and setting index and id on the result logs.